### PR TITLE
cosmosdbnosql: Added Cosmos DB NoSQL Semantic Cache Integration with tests and jupyter notebook

### PR DESCRIPTION
**Pull request by gsa9989**
* Added Cosmos DB NoSQL Semantic Cache Integration with tests and jupyter notebook

### CODE REVIEW

**Comment by gsa9989** [comment]
@ccurme can you please review this PR

File: libs/community/langchain_community/cache.py [hunk 1]
```diff
@@ -2275,6 +2280,101 @@ def _validate_enum_value(value: Any, enum_type: Type[Enum]) -> None:
...
+                    generations.extend(loads(document.metadata["return_val"]))
+                except Exception:
+                    logger.warning(
+                        "Retrieving a cache value that could not be deserialized "
+                        "properly. This is likely due to the cache being in an "
+                        "older format. Please recreate your cache to avoid this "
+                        "error."
+                    )
+                    # In a previous life we stored the raw text directly
+                    # in the table, so assume it's in that format.
```
Comment (ccurme) [inline libs/community/langchain_community/cache.py:2345]: This comment is mysterious (what table?)

File: libs/community/langchain_community/cache.py [hunk 2]
```diff
@@ -2275,6 +2280,101 @@ def _validate_enum_value(value: Any, enum_type: Type[Enum]) -> None:
...
+            "llm_string": llm_string,
+            "prompt": prompt,
+            "return_val": dumps([g for g in return_val]),
+        }
+        llm_cache.add_texts(texts=[prompt], metadatas=[metadata])
+
+    def clear(self, **kwargs: Any) -> None:
+        """Clear semantic cache for a given llm_string."""
+        database = self.cosmos_client.get_database_client(self.database_name)
+        container = database.get_container_client(self.container_name)
```
Comment (ccurme) [inline libs/community/langchain_community/cache.py]: removed `container` as it appeared unused

File: libs/community/tests/integration_tests/cache/test_azure_cosmosdbnosql_cache.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+        ]
+    }
+
+
+partition_key = PartitionKey(path="/id")
+cosmos_container_properties_test = {"partition_key": partition_key}
+cosmos_database_properties_test = {}
+
+
+# @pytest.fixture(scope="session")
```
Comment (ccurme) [inline libs/community/tests/integration_tests/cache/test_azure_cosmosdbnosql_cache.py:51]:
```suggestion
```

**Review by ccurme — COMMENTED** [review]
Thanks for the ping.

See our [contributing guidelines](https://python.langchain.com/v0.2/docs/contributing/code/setup/#formatting-and-linting) for how to format and lint code locally for future commits. I ran the formatter myself for now but there are still some type issues, would you mind taking a look? https://github.com/langchain-ai/langchain/actions/runs/10582689581/job/29323028501?pr=24424

---

The PR summary mentions a notebook, but I don't see any updates to documentation here. I think this would be good to add, otherwise the new cache will not be discoverable. If you edit `docs/docs/integrations/llm_caching.ipynb`, it will get built into the docs preview [here](https://langchain-git-fork-gsa9989-users-garagundicosm-e6d9e0-langchain.vercel.app/v0.2/docs/integrations/llm_caching/). Let me know if you have any questions!

File: libs/community/langchain_community/cache.py (hunk 1 above)
Comment (aayush3011) [reply libs/community/langchain_community/cache.py]: Removed.

File: libs/community/tests/integration_tests/cache/test_azure_cosmosdbnosql_cache.py (hunk 3 above)
Comment (aayush3011) [reply libs/community/tests/integration_tests/cache/test_azure_cosmosdbnosql_cache.py]: Removed

**Comment by aayush3011** [comment]
@ccurme Sorry, the PR through cracks. I have added the notebook please review.

**Comment by aayush3011** [comment]
Friendly ping @ccurme for a review here

**Comment by aayush3011** [comment]
Friendly ping @isahers1 @baskaryan for a review here

**Comment by aayush3011** [comment]
Friendly ping @ccurme @baskaryan @isahers1 for a review here.

File: docs/docs/integrations/llm_caching.ipynb [hunk 4]
```diff
@@ -1854,6 +1854,96 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "1e1cd93819921bf6",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "# The first time, it is not yet in cache, so it should take longer\n",
+    "llm.invoke(\"Tell me a joke\")"
```
Comment (ccurme) [inline docs/docs/integrations/llm_caching.ipynb:1932]: (nit) Could be worth running these cells to demonstrate the speedup.

File: libs/community/langchain_community/cache.py [hunk 5]
```diff
@@ -2102,7 +2106,7 @@ def __init__(
         ef_construction: int = 64,
         ef_search: int = 40,
         score_threshold: Optional[float] = None,
-        application_name: str = "LANGCHAIN_CACHING_PYTHON",
+        application_name: str = "LangChain-CDBNoSQL-SemanticCache-Python",
```
Comment (ccurme) [inline libs/community/langchain_community/cache.py:2109]: 1. Couldn't this break user workflows?
2. What is the motivation for this change? We have `AzureCosmosDBSemanticCache` and `AzureCosmosDBNoSqlSemanticCache` separately, but then we are specifying `NoSQL` here.

File: libs/community/langchain_community/vectorstores/azure_cosmos_db_no_sql.py [hunk 6]
```diff
@@ -12,7 +12,10 @@
 from langchain_community.vectorstores.utils import maximal_marginal_relevance
 
 if TYPE_CHECKING:
-    from azure.cosmos.cosmos_client import CosmosClient
+    from azure.cosmos import ContainerProxy, CosmosClient
+    from azure.identity import ClientSecretCredential
+
+USER_AGENT = ("LlamaIndex-CDBNoSql-VectorStore-Python",)
```
Comment (ccurme) [inline libs/community/langchain_community/vectorstores/azure_cosmos_db_no_sql.py:18]: LlamaIndex?

File: libs/community/langchain_community/vectorstores/azure_cosmos_db_no_sql.py [hunk 7]
```diff
@@ -274,38 +313,40 @@ def _similarity_search_with_score(
...
+            "c.id, c.text, c.metadata, c.embedding, "
+            "VectorDistance(c.embedding, @embeddings) AS SimilarityScore FROM c"
         )
 
         # Add where_clause if specified
         if pre_filter is not None and pre_filter.get("where_clause") is not None:
             query += " {}".format(pre_filter["where_clause"])
 
-        query += " ORDER BY VectorDistance(c.@embeddingKey, @embeddings)"
+        query += " ORDER BY VectorDistance(c.embedding, @embeddings)"
```
Comment (ccurme) [inline libs/community/langchain_community/vectorstores/azure_cosmos_db_no_sql.py:324]: Here too: is this a breaking change, and if so, what is the motivation?

Comment (aayush3011) [reply libs/community/langchain_community/vectorstores/azure_cosmos_db_no_sql.py:324]: This is not a breaking change. Sometimes the @embeddingKey is not working in the query and throwing a BadRequestException sometimes. And we are having it as c.embedding in our CosmosDB SDK as well

File: libs/community/langchain_community/cache.py (hunk 5 above)
Comment (aayush3011) [reply libs/community/langchain_community/cache.py:2109]: It's just a name change, this should not break any workflows. This user agent is being used for tracking purposes only. We want to have it in this pattern, as Azure CosmosDB is integrated in various AI frameworks (microsoft internal and external), and we want to create dashboards to track the usage using these user agents.

**Comment by aayush3011** [comment]
@ccurme Hey, apologies for the delay on the PR—it fell through the cracks. I've addressed all the comments, and it's ready for review. Please take a look when you have a chance. Thanks!

File: libs/community/langchain_community/cache.py (hunk 5 above)
Comment (aayush3011) [reply libs/community/langchain_community/cache.py]: Because we have seen that a lot of users are not specifying the application_name, and we can only see default value in our telemetry. So, that's the reason for this update as we can keep the application_name consistent across all our integrations.

**Comment by aayush3011** [comment]
Friendly ping @ccurme @baskaryan @isahers1 for a review here.

**Comment by aayush3011** [comment]
Friendly ping @efriis @ccurme @baskaryan for a final review here.

**Review by ccurme — APPROVED** [review]
